File: application/src/main/java/de/wuespace/telestion/project/daedalus2/messages/SystemT.java
package de.wuespace.telestion.project.daedalus2.messages;

import com.fasterxml.jackson.annotation.JsonProperty;
import de.wuespace.telestion.api.message.JsonMessage;
import de.wuespace.telestion.project.daedalus2.mavlink.SeedSystemT;

/**
 * A beautiful version of the Drehtest message
 *
 * @param timeLocal Seed local time
 */
public record SystemT(
		@JsonProperty
		long timeLocal,

		@JsonProperty
		Imu imu,

		@JsonProperty
		Baro baro,

		@JsonProperty
		float tachoRotRate,

		@JsonProperty
		ServoAmps servoAmps,

		@JsonProperty
		float batTemp,

		@JsonProperty
		float batHeatingAmps,

		@JsonProperty
		Volts volts,

		@JsonProperty
		Gps gps,

		@JsonProperty
		Filter filter,

		@JsonProperty
		Controller controller,

		@JsonProperty int currentState,

		@JsonProperty
		LidarServo lidarServo,

		@JsonProperty
		SystemConfig systemConfig,

		@JsonProperty
		AvailableStatus availableStatus
) implements JsonMessage {
	public SystemT() {
		this(
				System.currentTimeMillis(),
				new Imu(),
				new Baro(),
				0,
				new ServoAmps(),
				0,
				0,
				new Volts(),
				new Gps(),
				new Filter(),
				new Controller(),
				0,
				new LidarServo(),
				new SystemConfig(),
				new AvailableStatus()
		);
	}

	public SystemT(SeedSystemT rawSeedSystemT) {
		this(

				rawSeedSystemT.timeLocal(),
				new Imu(
						new Accelerometer(
								rawSeedSystemT.imuAccX(),
								rawSeedSystemT.imuAccY(),
								rawSeedSystemT.imuAccZ()
						),
						new Gyro(
								rawSeedSystemT.imuGyroX(),
								rawSeedSystemT.imuGyroY(),
								rawSeedSystemT.imuGyroZ()
						)
				),
				new Baro(rawSeedSystemT.baroPress(), rawSeedSystemT.baroTemp(), rawSeedSystemT.vacuumBaroPress()),
				rawSeedSystemT.tachoRotRate(),
				new ServoAmps(
						rawSeedSystemT.swashplateServo1Amps(),
						rawSeedSystemT.swashplateServo2Amps(),
						rawSeedSystemT.swashplateServo3Amps(),
						rawSeedSystemT.finServoAmps()
				),
				rawSeedSystemT.batTemp(),
				rawSeedSystemT.batHeatingAmps(),
				new Volts(
						rawSeedSystemT.rxsmVolts(),
						rawSeedSystemT.bat1Volts(),
						rawSeedSystemT.bat2Volts(),
						rawSeedSystemT.rail3V3Volts()
				),
				new Gps(
						rawSeedSystemT.gpsLat(),
						rawSeedSystemT.gpsLong(),
						rawSeedSystemT.gpsHdop(),
						rawSeedSystemT.gpsAlt(),
						rawSeedSystemT.gpsQuality(),
						rawSeedSystemT.gpsSatsused()
				),
				new Filter(
						rawSeedSystemT.filterVelVertical(),
						rawSeedSystemT.filterVelVerticalInd(),
						rawSeedSystemT.filterHeightGround(),
						rawSeedSystemT.filterRotorRotRate(),
						rawSeedSystemT.fiterBodyRotRate()
				),
				new Controller(
						rawSeedSystemT.controllerId(),
						rawSeedSystemT.controllerBladePitch(),
						rawSeedSystemT.controllerFinAngle()
				),
				rawSeedSystemT.stateCur(),
				new LidarServo(rawSeedSystemT.lidarServo()),
				new SystemConfig(rawSeedSystemT.systemConfig()),
				new AvailableStatus(rawSeedSystemT.availableStatus())
		);
	}
}

File: client/src/model/mongo-collections/system-t.ts
import { MongoCollection } from './base';
import { JavaMessage } from '../java-message';

export interface Accelerometer extends JavaMessage {
	x: number;
	y: number;
	z: number;
	className: 'de.wuespace.telestion.project.daedalus2.messages.Accelerometer';
}

export interface Gyroscope extends JavaMessage {
	x: number;
	y: number;
	z: number;
	className: 'de.wuespace.telestion.project.daedalus2.messages.Gyro';
}

export interface IMU extends JavaMessage {
	acc: Accelerometer;
	gyro: Gyroscope;
	className: 'de.wuespace.telestion.project.daedalus2.messages.Imu';
}

export interface Barometer extends JavaMessage {
	press: number;
	temp: number;
	vacuumPress: number;
	className: 'de.wuespace.telestion.project.daedalus2.messages.Baro';
}

export interface ServoAmps extends JavaMessage {
	swashplate_servo1_amps: number;
	swashplate_servo2_amps: number;
	swashplate_servo3_amps: number;
	fin_servo_amps: number;
	className: 'de.wuespace.telestion.project.daedalus2.messages.ServoAmps';
}

export interface Volts extends JavaMessage {
	rxsmVolts: number;
	bat1_volts: number;
	bat2_volts: number;
	rail3V3Volts: number;
	className: 'de.wuespace.telestion.project.daedalus2.messages.Volts';
}

export interface GPS extends JavaMessage {
	latitude: number;
	longitude: number;
	horizontalDilutionOfPrecision: number;
	altitude: number;
	quality: number;
	numberOfSatsUsed: number;
	className: 'de.wuespace.telestion.project.daedalus2.messages.Gps';
}

export interface Filter extends JavaMessage {
	velVertical: number;
	velVerticalInd: number;
	heightGround: number;
	rotorRotRate: number;
	bodyRotRate: number;
	className: 'de.wuespace.telestion.project.daedalus2.messages.Filter';
}

export interface Controller extends JavaMessage {
	id: number;
	bladePitch: number;
	finAngle: number;
	className: 'de.wuespace.telestion.project.daedalus2.messages.Controller';
}

export interface LidarServo extends JavaMessage {
	lidarServoOn: boolean;
	lidarServoOpen: boolean;
	className: 'de.wuespace.telestion.project.daedalus2.messages.LidarServo';
}

export interface SystemConfig extends JavaMessage {
	rxsmAllowed: boolean;
	bat1Allowed: boolean;
	bat2Allowed: boolean;
	className: 'de.wuespace.telestion.project.daedalus2.messages.SystemConfig';
}

export interface AvailableStatus extends JavaMessage {
	imuAccAvail: boolean;
	imuGyroAvail: boolean;
	baroAvail: boolean;
	vacuumBaroAvail: boolean;
	tachoRotAvail: boolean;
	servoAmpsAvail: boolean;
	batTempAvail: boolean;
	voltsAvail: boolean;
	className: 'de.wuespace.telestion.project.daedalus2.messages.AvailableStatus';
}

export interface SystemT extends MongoCollection {
	timeLocal: number;
	imu: IMU;
	baro: Barometer;
	tachoRotRate: number;
	servoAmps: ServoAmps;
	batTemp: number;
	batHeatingAmps: number;
	volts: Volts;
	gps: GPS;
	filter: Filter;
	Controller: Controller;
	currentState: number;
	lidarServo: LidarServo;
	systemConfig: SystemConfig;
	availableStatus: AvailableStatus;
	className: 'de.wuespace.telestion.project.daedalus2.messages.SystemT';
	datetime: {
		$date: string;
	};
}
